crates/solver/src/settlement_access_list.rs

```diff
@@ -1,37 +1,123 @@
+
 #[async_trait::async_trait]
 pub trait AccessListEstimating: Send + Sync {
     async fn estimate_access_list(
         &self,
         tx: &TransactionBuilder<DynTransport>,
+        partial_access_list: Option<AccessList>,
     ) -> Result<AccessList> {
-        self.estimate_access_lists(std::slice::from_ref(tx))
+        self.estimate_access_lists(std::slice::from_ref(tx), partial_access_list)
             .await?
             .into_iter()
             .next()
             .unwrap()
     }
-    /// The function guarantees the same size and order of input and output containers
+
+    // TODO This method should not be needed anymore, refactor it in a follow-up.
+    /// Guarantees the same length and order of input and output values.
     async fn estimate_access_lists(
         &self,
         txs: &[TransactionBuilder<DynTransport>],
+        partial_access_list: Option<AccessList>,
     ) -> Result<Vec<Result<AccessList>>>;
 }
 
+/// Does access list estimation for a transaction interacting with our settlement contract. In
+/// particular, our settlement contract will fail if the receiver is a smart contract. Because of
+/// this, if the receiver is a smart contract and we try to estimate the access list, the access
+/// list estimation will also fail.
+///
+/// The reason why this failure happens is because the Ethereum protocol sets a hard gas limit on
+/// transferring ETH into a smart contract, which some contracts exceed unless the access list is
+/// already specified.
+///
+/// The solution is to do access list estimation in two steps: first, simulate moving 1 wei
+/// into every smart contract to get a partial access list, and then simulate the full access list,
+/// passing the partial access list into the simulation. This way the settlement contract does not
+/// fail, and hence the full access list estimation also does not fail.
+pub async fn estimate_settlement_access_list(
+    estimating: &dyn AccessListEstimating,
+    web3: Web3,
+    solver_account: Account,
+    settlement: &Settlement,
+    tx: &TransactionBuilder<DynTransport>,
+) -> Result<AccessList> {
+    // Generate partial access lists for all smart contracts
+    let partial_access_lists = join_all(settlement.trades().map(|trade| async {
+        let buy_token = trade.order.data.buy_token;
+        let owner = trade.order.metadata.owner;
+        let partial_access_list =
+            if buy_token == BUY_ETH_ADDRESS && is_smart_contract(&web3, owner).await? {
+                Some(
+                    estimating
+                        .estimate_access_list(
+                            &TransactionBuilder::new(web3.clone())
+                                .from(solver_account.clone())
+                                .to(owner)
+                                .value(1.into()),
+                            None,
+                        )
+                        .await?,
+                )
+            } else {
+                None
```

Comment:
Not sure but maybe the code might look a bit nicer if this future returned `Result<AccessList>` instead of `Result<Option<AccessList>>`. This assumes that returning an empty `AccessList` instead of `None` doesn't cause problems while merging access lists.